File: Raven.Abstractions/Json/Linq/DictionaryWithParentSnapshot.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;
using Raven.Imports.Newtonsoft.Json.Linq;
using System.Linq;

namespace Raven.Json.Linq
{
	internal class DictionaryWithParentSnapshot : IDictionary<string, RavenJToken>
	{
		private readonly IEqualityComparer<string> comparer;
		private static readonly RavenJToken DeletedMarker = new RavenJValue("*DeletedMarker*", JTokenType.Null);

		private readonly DictionaryWithParentSnapshot parentSnapshot;
		private int count;
		private IDictionary<string, RavenJToken> localChanges;
		private string snapshotMsg;

		protected IDictionary<string, RavenJToken> LocalChanges
		{
			get
			{
				if (localChanges == null)
					localChanges = new Dictionary<string, RavenJToken>(comparer);
				return localChanges;
			}
		}

		public DictionaryWithParentSnapshot(IEqualityComparer<string> comparer)
		{
			this.comparer = comparer;
		}

		private DictionaryWithParentSnapshot(DictionaryWithParentSnapshot previous)
		{
			comparer = previous.comparer;
			parentSnapshot = previous;
		}

		#region Dictionary<string,TValue> Members

		public void Add(string key, RavenJToken value)
		{
			Debug.Assert(!String.IsNullOrWhiteSpace(key),"key must _never_ be null/empty/whitespace");

			if (IsSnapshot)
				throw new InvalidOperationException(snapshotMsg ?? "Cannot modify a snapshot, this is probably a bug");

			if (ContainsKey(key))
				throw new ArgumentException(string.Format("An item with the same key has already been added: '{0}'", key));

			count += 1;
			LocalChanges[key] = value; // we can't use Add, because LocalChanges may contain a DeletedMarker
		}

		public bool ContainsKey(string key)
		{
			RavenJToken token;
			if (localChanges != null && localChanges.TryGetValue(key, out token))
			{
				if (token == DeletedMarker)
					return false;
				return true;
			}
			return (parentSnapshot != null && parentSnapshot.TryGetValue(key, out token) && token != DeletedMarker);
		}

		public ICollection<string> Keys
		{
			get
			{
				if (localChanges == null)
				{
					if (parentSnapshot != null)
					{
						return parentSnapshot.Keys;
					}
					return new HashSet<string>();
				}

				ICollection<string> ret = new HashSet<string>();
				if (parentSnapshot != null)
				{
					foreach (var key in parentSnapshot.Keys)
					{
						if (LocalChanges.ContainsKey(key))
							continue;
						ret.Add(key);
					}
				}

				foreach (var key in LocalChanges.Keys)
				{
					RavenJToken value;
					if (LocalChanges.TryGetValue(key, out value) == false ||
						value == DeletedMarker)
						continue;
					ret.Add(key);
				}

				return ret;
			}
		}

		public bool Remove(string key)
		{
			if (IsSnapshot)
				throw new InvalidOperationException("Cannot modify a snapshot, this is probably a bug");

			RavenJToken parentToken = null;

			bool parentHasIt = parentSnapshot != null &&
							   parentSnapshot.TryGetValue(key, out parentToken);

			RavenJToken token;
			if (LocalChanges.TryGetValue(key, out token) == false)
			{
				if (parentHasIt && parentToken != DeletedMarker)
				{
					LocalChanges[key] = DeletedMarker; 
					count -= 1;
					return true;
				}
				return false;
			}
			if (token == DeletedMarker)
				return false;
			count -= 1;
			LocalChanges[key] = DeletedMarker;
			return true;
		}

		public bool TryGetValue(string key, out RavenJToken value)
		{			
			value = null;
			RavenJToken unsafeVal;
			if (localChanges != null && localChanges.TryGetValue(key, out unsafeVal))
			{
				if (unsafeVal == DeletedMarker)
					return false;

				value = unsafeVal;				
				return true;
			}

			if (parentSnapshot == null ||
				!parentSnapshot.TryGetValue(key, out unsafeVal) ||
				unsafeVal == DeletedMarker)
				return false;

			if (IsSnapshot == false && unsafeVal != null)
			{
				if (unsafeVal.IsSnapshot == false && unsafeVal.Type != JTokenType.Object)
                    unsafeVal.EnsureCannotBeChangeAndEnableSnapshotting();
			}

			value = unsafeVal;
			return true;
		}

		public ICollection<RavenJToken> Values
		{
			get
			{
				ICollection<RavenJToken> ret = new HashSet<RavenJToken>();
				foreach (var key in Keys)
				{
					ret.Add(this[key]);
				}
				return ret;
			}
		}

		public RavenJToken this[string key]
		{
			get
			{			
				RavenJToken token;
				if (TryGetValue(key, out token))
					return token;
				throw new KeyNotFoundException(key);
			}
			set
			{
				if (IsSnapshot)
					throw new InvalidOperationException("Cannot modify a snapshot, this is probably a bug");
				var isInsert = ContainsKey(key) == false;
				LocalChanges[key] = value;
				if (isInsert)
					count += 1;
			}
		}

		#endregion

		public IEnumerator<KeyValuePair<string, RavenJToken>> GetEnumerator()
		{
			if (parentSnapshot != null)
			{
				foreach (var item in parentSnapshot)
				{
					if (item.Key == null)
						continue;
					if (LocalChanges.ContainsKey(item.Key))
						continue;
					yield return item;
				}
			}
			foreach (var localChange in LocalChanges)
			{
				if (localChange.Value == DeletedMarker)
					continue;
				yield return localChange;
			}

		}

		IEnumerator IEnumerable.GetEnumerator()
		{
			return GetEnumerator();
		}

		public void Add(KeyValuePair<string, RavenJToken> item)
		{
			Add(item.Key, item.Value);
		}

		public void Clear()
		{
			// we either already have count set to -1, or it will be invalidated by a call to Remove below

			foreach (var key in Keys.ToArray()) // clone the values for the iteration
			{
				Remove(key);
			}
		}

		public bool Contains(KeyValuePair<string, RavenJToken> item)
		{
			throw new NotImplementedException();
		}

		public void CopyTo(KeyValuePair<string, RavenJToken>[] array, int arrayIndex)
		{
			if (parentSnapshot != null)
			{
				parentSnapshot.CopyTo(array, arrayIndex);
				arrayIndex += parentSnapshot.Count;
			}
			LocalChanges.CopyTo(array, arrayIndex);
		}

		public bool Remove(KeyValuePair<string, RavenJToken> item)
		{
			throw new NotImplementedException();
		}

		public int Count
		{
			get
			{
				if (parentSnapshot != null)
					return count + parentSnapshot.Count;
				return count;
			}
		}

		public bool IsReadOnly
		{
			get { return false; }
		}

		public bool CaseInsensitivePropertyNames { get; set; }
		public bool IsSnapshot { get; private set; }

		public DictionaryWithParentSnapshot CreateSnapshot()
		{
			if (IsSnapshot == false)
				throw new InvalidOperationException("Cannot create snapshot without previously calling EnsureSnapShot");
			return new DictionaryWithParentSnapshot(this);
		}

		public void EnsureSnapshot(string msg = null)
		{
			snapshotMsg = msg;
			IsSnapshot = true;
		}
	}
}

File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/json/linq/DictionaryWithParentSnapshot.java
package net.ravendb.abstractions.json.linq;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;

import net.ravendb.abstractions.basic.Reference;


public class DictionaryWithParentSnapshot implements Map<String, RavenJToken>, Iterable<Map.Entry<String, RavenJToken>> {

  private final Comparator<String> comparer;

  private static final RavenJToken DELETED_MARKER = new RavenJValue("*DeletedMarker*", JTokenType.NULL);

  private final DictionaryWithParentSnapshot parentSnapshot;
  private int count;
  private Map<String, RavenJToken> localChanges;
  private String snapshotMsg;
  private boolean snapshot;

  public Map<String, RavenJToken> getLocalChanges() {
    if (localChanges == null) {
      localChanges = new TreeMap<> (comparer);
    }
    return localChanges;
  }

  public DictionaryWithParentSnapshot(Comparator<String> comparer) {
    this.comparer = comparer;
    this.parentSnapshot = null;
  }


  public DictionaryWithParentSnapshot(DictionaryWithParentSnapshot previous) {
    this.parentSnapshot = previous;
    this.comparer = previous.comparer;
  }

  /* (non-Javadoc)
   * @see java.util.HashMap#put(java.lang.Object, java.lang.Object)
   */
  @Override
  public RavenJToken put(String key, RavenJToken value) {
    if (isSnapshot()) {
      throw new IllegalStateException(snapshotMsg != null ? snapshotMsg
        : "Cannot modify a snapshot, this is probably a bug");
    }
    if (!containsKey(key)) {
      count++;
    }
    getLocalChanges().put(key, value);
    return value;
  }

  /* (non-Javadoc)
   * @see java.util.Map#containsKey(java.lang.Object)
   */
  @Override
  public boolean containsKey(Object keyObject) {
    String key = (String) keyObject;
    RavenJToken token;
    if (getLocalChanges().containsKey(key)) {
      token = getLocalChanges().get(key);
      if (token == DELETED_MARKER) {
        return false;
      }
      return true;
    }
    return (parentSnapshot != null && parentSnapshot.containsKey(key) && parentSnapshot.get(key) != DELETED_MARKER);
  }

  /* (non-Javadoc)
   * @see java.util.Map#keySet()
   */
  @Override
  public Set<String> keySet() {
    if (getLocalChanges().isEmpty()) {
      if (parentSnapshot != null) {
        return parentSnapshot.keySet();
      }
      return new HashSet<>();
    }

    Set<String> ret = new HashSet<>();
    if (parentSnapshot != null) {
      for (String key : parentSnapshot.keySet()) {
        if (!getLocalChanges().containsKey(key)) {
          ret.add(key);
        }
      }
    }
    for (String key : getLocalChanges().keySet()) {
      if (getLocalChanges().containsKey(key) && getLocalChanges().get(key) != DELETED_MARKER) {
        ret.add(key);
      }
    }

    return ret;
  }

  /* (non-Javadoc)
   * @see java.util.Map#remove(java.lang.Object)
   */
  @Override
  public RavenJToken remove(Object keyObject) {
    String key = (String) keyObject;
    if (isSnapshot()) {
      throw new IllegalStateException("Cannot modify a snapshot, this is probably a bug");
    }
    boolean parentHasIt = false;
    RavenJToken parentToken = null;

    if (parentSnapshot != null) {
      parentToken = parentSnapshot.get(key);
      if (parentToken != null) {
        parentHasIt = true;
      }
    }

    if (!getLocalChanges().containsKey(key)) {
      if (parentHasIt && parentToken != DELETED_MARKER) {
        getLocalChanges().put(key, DELETED_MARKER);
        count--;
        return parentToken;
      }
      return null;
    }
    RavenJToken token = getLocalChanges().get(key);
    if (token == DELETED_MARKER) {
      return null;
    }
    count--;
    getLocalChanges().put(key, DELETED_MARKER);
    return token;
  }

  /* (non-Javadoc)
   * @see java.util.Map#get(java.lang.Object)
   */
  @Override
  public RavenJToken get(Object keyObject) {
    String key = (String) keyObject;
    if (getLocalChanges().containsKey(key)) {
      RavenJToken unsafeVal = getLocalChanges().get(key);
      if (unsafeVal == DELETED_MARKER) {
        return null;
      }
      return unsafeVal;
    }
    if (parentSnapshot == null || !parentSnapshot.containsKey(key) || parentSnapshot.get(key) == DELETED_MARKER) {
      return null;
    }
    RavenJToken unsafeVal = parentSnapshot.get(key);
    if (!isSnapshot() && unsafeVal != null) {
      if (!unsafeVal.isSnapshot() && unsafeVal.getType() != JTokenType.OBJECT) {
        unsafeVal.ensureCannotBeChangeAndEnableShapshotting();
      }
    }
    return unsafeVal;
  }

  /* (non-Javadoc)
   * @see java.util.Map#values()
   */
  @Override
  public Collection<RavenJToken> values() {
    Collection<RavenJToken> tokens = new ArrayList<>();
    for (String key : keySet()) {
      tokens.add(get(key));
    }
    return tokens;
  }

  public boolean isSnapshot() {
    return snapshot;
  }

  public DictionaryWithParentSnapshot createSnapshot() {
    if (!isSnapshot()) {
      throw new IllegalStateException("Cannot create snapshot without previously calling EnsureSnapshot");
    }
    return new DictionaryWithParentSnapshot(this);
  }

  /* (non-Javadoc)
   * @see java.util.Map#clear()
   */
  @Override
  public void clear() {
    for (String key : keySet()) {
      remove(key);
    }
  }

  /* (non-Javadoc)
   * @see java.util.Map#isEmpty()
   */
  @Override
  public boolean isEmpty() {
    return size() == 0;
  }

  /* (non-Javadoc)
   * @see java.util.Map#putAll(java.util.Map)
   */
  @Override
  public void putAll(Map<? extends String, ? extends RavenJToken> m) {
    for (String key : m.keySet()) {
      put(key, m.get(key));
    }
  }

  /* (non-Javadoc)
   * @see java.util.Map#containsValue(java.lang.Object)
   */
  @Override
  public boolean containsValue(Object value) {
    throw new IllegalStateException("Not implemeneted");
  }

  /* (non-Javadoc)
   * @see java.util.Map#entrySet()
   */
  @Override
  public Set<Entry<String, RavenJToken>> entrySet() {
    Set<Entry<String, RavenJToken>> entries = new HashSet<>();
    Iterator<java.util.Map.Entry<String, RavenJToken>> iterator = iterator();
    while (iterator.hasNext()) {
      entries.add(iterator.next());
    }
    return entries;
  }

  /* (non-Javadoc)
   * @see java.util.Map#size()
   */
  @Override
  public int size() {
    if (parentSnapshot != null) {
      return count + parentSnapshot.size();
    }
    return count;
  }

  public void ensureSnapshot() {
    ensureSnapshot(null);
  }

  public void ensureSnapshot(String msg) {
    snapshot = true;
    snapshotMsg = msg;
  }

  @Override
  public Iterator<java.util.Map.Entry<String, RavenJToken>> iterator() {
    return new DictionaryInterator();
  }

  public boolean tryGetValue(String key, Reference<RavenJToken> value) {
    value.value = null;
    Reference<RavenJToken> unsafeVal = new Reference<>();
    if (getLocalChanges() != null && getLocalChanges().containsKey(key)) {
      unsafeVal.value = getLocalChanges().get(key);
      if (DELETED_MARKER.equals(unsafeVal.value)) return false;

      value.value = unsafeVal.value;
      return true;
    }

    if (parentSnapshot == null ||
      !parentSnapshot.tryGetValue(key, unsafeVal) ||
      DELETED_MARKER.equals(unsafeVal.value)) return false;

    if (!isSnapshot() && unsafeVal.value != null) {
      if (unsafeVal.value.isSnapshot() && JTokenType.OBJECT.equals(unsafeVal.value.getType()))
        unsafeVal.value.ensureCannotBeChangeAndEnableShapshotting();
    }

    value.value = unsafeVal.value;

    return true;
  }

  class DictionaryInterator implements Iterator<java.util.Map.Entry<String, RavenJToken>> {

    private boolean parentProcessed = false;
    private Iterator<java.util.Map.Entry<String, RavenJToken>> parentIterator;
    private Iterator<java.util.Map.Entry<String, RavenJToken>> localIterator;

    private Iterator<java.util.Map.Entry<String, RavenJToken>> getCurrentIterator() {
      if (!parentProcessed) {
        if (parentSnapshot != null) {
          if (parentIterator == null) {
            parentIterator = parentSnapshot.iterator();
          }
          if (parentIterator.hasNext()) {
            return parentIterator;
          } else {
            //no more elements in parent iterator
            parentProcessed = true;
          }
        } else {
          parentProcessed = true;
        }
      }
      if (localIterator == null) {
        Map<String, RavenJToken> entrySetMap = new HashMap<>();
        for (Map.Entry<String, RavenJToken> entry: getLocalChanges().entrySet()) {
          if (entry.getValue() != DELETED_MARKER) {
            entrySetMap.put(entry.getKey(), entry.getValue());
          }
        }
        localIterator = entrySetMap.entrySet().iterator();

      }
      return localIterator;
    }

    @Override
    public boolean hasNext() {
      return getCurrentIterator().hasNext();
    }

    @Override
    public java.util.Map.Entry<String, RavenJToken> next() {
      return getCurrentIterator().next();
    }

    @Override
    public void remove() {
      throw new IllegalStateException("Deleting elements in iterator is not implemneted!");
    }

  }

}
